### PR TITLE
Rewrite registries package

### PR DESCRIPTION
**Pull request by bwateratmsft**

### CODE REVIEW

File: packages/vscode-docker-registries/src/clients/GenericRegistryV2/GenericRegistryV2DataProvider.ts [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        const index = trackedRegistryStrings.findIndex(r => r === registryUriString);
+        if (index !== -1) {
+            trackedRegistryStrings.splice(index, 1);
+            void this.extensionContext.globalState.update(TrackedRegistriesKey, trackedRegistryStrings);
+        }
+
+        // remove credentials from auth provider
+        await this.authenticationProviders.get(registryUriString)?.removeSession();
+        this.authenticationProviders.delete(registryUriString);
+        // TODO: check if the map of auth providers is empty, if so, remove the root from the tree
```
Comment (bwateratmsft) [inline packages/vscode-docker-registries/src/clients/GenericRegistryV2/GenericRegistryV2DataProvider.ts:151]: Do?

File: packages/vscode-docker-registries/src/clients/GenericRegistryV2/GenericRegistryV2DataProvider.ts [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    readonly additionalContextValues: ['genericRegistryV2Registry'];
+}
+
+interface GenericV2RegistryTag extends V2Tag {
+    readonly additionalContextValues: ['genericRegistryV2Tag'];
+}
+
+export function isGenericV2Registry(item: unknown): item is GenericV2Registry {
+    return isContextValueRegistryItem(item) && item.additionalContextValues?.includes('genericRegistryV2Registry') === true;
+}
```
Comment (bwateratmsft) [inline packages/vscode-docker-registries/src/clients/GenericRegistryV2/GenericRegistryV2DataProvider.ts:31]: We can maybe replace all this with overriding `getChildren` and attaching a `generic` context value

File: packages/vscode-docker-registries/src/clients/GitHub/GitHubRegistryDataProvider.ts [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+                parent: root,
+                baseUrl: GitHubContainerRegistryUri,
+                label: org,
+                type: 'commonregistry',
+                additionalContextValues: ['githubRegistry']
+            }
+        ));
+    }
+
+    // TODO: GitHub's catalog endpoint uses standard paging, so this could be simplified to call super.getRepositories with just the added query parameter for the last repository name
```
Comment (bwateratmsft) [inline packages/vscode-docker-registries/src/clients/GitHub/GitHubRegistryDataProvider.ts:87]: Do

File: packages/vscode-docker-registries/src/clients/GenericRegistryV2/GenericRegistryV2DataProvider.ts [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+    public async getTags(repository: V2Repository): Promise<GenericV2RegistryTag[]> {
+        const tags = await super.getTags(repository);
+        const tagsWithAdditionalContext: GenericV2RegistryTag[] = tags.map(tag => ({
+            ...tag,
+            additionalContextValues: ['genericRegistryV2Tag']
+        }));
+
+        return tagsWithAdditionalContext;
+    }
```
Comment (bwateratmsft) [inline packages/vscode-docker-registries/src/clients/GenericRegistryV2/GenericRegistryV2DataProvider.ts:82]: What purpose does this additional context value serve?

File: packages/vscode-docker-registries/src/clients/GitHub/GitHubRegistryDataProvider.ts [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+
+    public async getTags(repository: V2Repository): Promise<V2Tag[]> {
+        const tags = await super.getTags(repository);
+        const tagsWithAdditionalContext: V2Tag[] = tags.map(tag => ({
+            ...tag,
+            additionalContextValues: ['githubRegistryTag']
+        }));
+
+        return tagsWithAdditionalContext;
+    }
```
Comment (bwateratmsft) [inline packages/vscode-docker-registries/src/clients/GitHub/GitHubRegistryDataProvider.ts:137]: What purpose does this additional context value serve?

File: packages/vscode-docker-registries/src/clients/GitHub/GitHubRegistryDataProvider.ts [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+
+    protected getAuthenticationProvider(item: V2RegistryItem): AuthenticationProvider<never> {
+        return this.authenticationProvider;
+    }
+
+    protected override async getTagDetails(repository: V2Repository, tag: string): Promise<Date | undefined> {
```
Comment (bwateratmsft) [inline packages/vscode-docker-registries/src/clients/GitHub/GitHubRegistryDataProvider.ts:143]: Method name is strange for this

File: packages/vscode-docker-registries/src/clients/RegistryV2/RegistryV2DataProvider.ts [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+        const digest = response.headers['docker-content-digest'];
+        if (!digest) {
+            throw new Error('Could not find digest');
+        }
+
+        return digest;
+    }
+
+    protected async getTagDetails(repository: V2Repository, tag: string): Promise<Date | undefined> {
```
Comment (bwateratmsft) [inline packages/vscode-docker-registries/src/clients/RegistryV2/RegistryV2DataProvider.ts:134]: Weird method name for this

File: packages/vscode-docker-registries/src/wizard/showInputBox.ts [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+                    reject(new CancellationError());
+                }),
+            );
+            inputBox.show();
+        });
+    } finally {
+        disposables.forEach(d => {
+            d.dispose();
+        });
+    }
```
Comment (bwateratmsft) [inline packages/vscode-docker-registries/src/wizard/showInputBox.ts:58]: This implementation is complex, perhaps more than it's worth.

File: packages/vscode-docker-registries/src/clients/GenericRegistryV2/GenericRegistryV2DataProvider.ts (hunk 1 above)
Comment (alexyaang) [reply packages/vscode-docker-registries/src/clients/GenericRegistryV2/GenericRegistryV2DataProvider.ts:151]: mhmm not sure if this is worth, what do you think?

File: packages/vscode-docker-registries/src/clients/GenericRegistryV2/GenericRegistryV2DataProvider.ts (hunk 2 above)
Comment (alexyaang) [reply packages/vscode-docker-registries/src/clients/GenericRegistryV2/GenericRegistryV2DataProvider.ts:31]: that sounds good to me

File: packages/vscode-docker-registries/src/clients/GenericRegistryV2/GenericRegistryV2DataProvider.ts (hunk 4 above)
Comment (alexyaang) [reply packages/vscode-docker-registries/src/clients/GenericRegistryV2/GenericRegistryV2DataProvider.ts:82]: After we make the context value change in delete image, this is no longer needed

File: packages/vscode-docker-registries/src/wizard/showInputBox.ts (hunk 8 above)
Comment (alexyaang) [reply packages/vscode-docker-registries/src/wizard/showInputBox.ts:58]: which part are you referring to? the validation logic?

File: packages/vscode-docker-registries/src/clients/GenericRegistryV2/GenericRegistryV2DataProvider.ts (hunk 1 above)
Comment (bwateratmsft) [reply packages/vscode-docker-registries/src/clients/GenericRegistryV2/GenericRegistryV2DataProvider.ts:151]: Well, is there any way to make the node disappear after a disconnect?

File: packages/vscode-docker-registries/src/wizard/showInputBox.ts (hunk 8 above)
Comment (bwateratmsft) [reply packages/vscode-docker-registries/src/wizard/showInputBox.ts:58]: The whole thing really. You could get almost the same behavior with:
```typescript
    return await window.showInputBox({
        prompt: options.prompt,
        placeHolder: options.placeholder,
        ignoreFocusOut: true,
        password: !!options.isSecretStep,
        validateInput: options.validateInput,
    });
```

File: packages/vscode-docker-registries/src/clients/GenericRegistryV2/GenericRegistryV2DataProvider.ts (hunk 1 above)
Comment (alexyaang) [reply packages/vscode-docker-registries/src/clients/GenericRegistryV2/GenericRegistryV2DataProvider.ts:150]: no, not currently

File: packages/vscode-docker-registries/src/wizard/showInputBox.ts (hunk 8 above)
Comment (alexyaang) [reply packages/vscode-docker-registries/src/wizard/showInputBox.ts:58]: you're right... looks like i complicated things by a lot

File: packages/vscode-docker-registries/src/wizard/showInputBox.ts [hunk 9]
```diff
@@ -3,57 +3,17 @@
...
-            );
-            inputBox.show();
-        });
-    } finally {
-        disposables.forEach(d => {
-            d.dispose();
-        });
-    }
+export async function showInputBox(options: RegistryWizardPromptStepOptions): Promise<string | undefined> {
+    return await window.showInputBox({
```
Comment (bwateratmsft) [inline packages/vscode-docker-registries/src/wizard/showInputBox.ts:12]: I like the idea of reinterpreting `undefined` to throw a `CancellationError` though, we should keep that part.

Comment (alexyaang) [reply packages/vscode-docker-registries/src/wizard/showInputBox.ts:12]: Yeah I agree. I opened a PR here https://github.com/microsoft/vscode-docker-extensibility/pull/188/files, should simplify it by a lot

File: packages/vscode-docker-registries/src/auth/DockerHubAuthProvider.ts [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+import * as vscode from 'vscode';
+import { AuthenticationProvider } from '../contracts/AuthenticationProvider';
+import { LoginInformation } from '../contracts/BasicCredentials';
+import { DockerHubRegistryUrl, DockerHubRequestUrl } from '../clients/DockerHub/DockerHubRegistryDataProvider';
+import { httpRequest } from '../utils/httpRequest';
+import { BasicAuthProvider } from './BasicAuthProvider';
+
+export class DockerHubAuthProvider extends BasicAuthProvider implements AuthenticationProvider {
+    // TODO: this token expires after a month, should we bother to actually refresh it?
```
Comment (alexyaang) [inline packages/vscode-docker-registries/src/auth/DockerHubAuthProvider.ts:14]: Do?

Comment (bwateratmsft) [reply packages/vscode-docker-registries/src/auth/DockerHubAuthProvider.ts:14]: Nahh, who keeps VSCode running for a month?

File: packages/vscode-docker-registries/src/clients/DockerHub/DockerHubRegistryDataProvider.ts [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+
+        const sortedOrgsAndNamespaces = Array.from(orgsAndNamespaces).sort();
+
+        for (const orgOrNamespace of sortedOrgsAndNamespaces) {
+            results.push({
+                parent: root,
+                label: orgOrNamespace,
+                type: 'commonregistry',
+                additionalContextValues: ['dockerHubRegistry'],
+                baseUrl: DockerHubRegistryUrl.with({ path: orgOrNamespace }),
```
Comment (alexyaang) [inline packages/vscode-docker-registries/src/clients/DockerHub/DockerHubRegistryDataProvider.ts:99]: TODO: change this, baseUrl should just be `DockerHubRegistryUrl`

File: packages/vscode-docker-registries/src/clients/GenericRegistryV2/GenericRegistryV2DataProvider.ts (hunk 4 above)
Comment (alexyaang) [reply packages/vscode-docker-registries/src/clients/GenericRegistryV2/GenericRegistryV2DataProvider.ts:82]: https://github.com/microsoft/vscode-docker/pull/4026#discussion_r1313331522 should simplify this as well

File: packages/vscode-docker-registries/src/clients/GenericRegistryV2/GenericRegistryV2DataProvider.ts (hunk 1 above)
Comment (alexyaang) [reply packages/vscode-docker-registries/src/clients/GenericRegistryV2/GenericRegistryV2DataProvider.ts:136]: we'll do this as a separate PR, issue is tracked here https://github.com/microsoft/vscode-docker-extensibility/issues/193

**Review by alexyaang — APPROVED** [review]

**Pull request merged**